File: android/data/src/main/java/com/ex/lolcompose/data/repository/ChampionRepositoryImpl.kt
package com.ex.lolcompose.data.repository

import com.ex.lolcompose.data.datasource.ChampionRemoteDataSource
import com.ex.lolcompose.domain.common.DataResult
import com.ex.lolcompose.domain.model.Champion
import com.ex.lolcompose.domain.repository.ChampionRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.flow
import javax.inject.Inject

class ChampionRepositoryImpl @Inject constructor(
    private val remoteDataSource: ChampionRemoteDataSource
) : ChampionRepository {

    override fun getChampionList(): Flow<DataResult<List<Champion>>> = flow<DataResult<List<Champion>>> {
        val champions = remoteDataSource.getChampionList()
            .data
            .values
            .sortedBy { it.name }
        emit(DataResult.Success(champions))
    }.catch { exception ->
        emit(DataResult.Error(exception))
    }

    override fun getChampion(id: String): Flow<DataResult<Champion>> = flow<DataResult<Champion>> {
        val champion = remoteDataSource.getChampion(id)
            .data[id]
            ?: throw NoSuchElementException("Champion not found: $id")
        emit(DataResult.Success(champion))
    }.catch { exception ->
        emit(DataResult.Error(exception))
    }
}

File: flutter/lib/features/champion/data/repository/champion_repository_impl.dart
import 'package:dio/dio.dart';
import 'package:lol_champion/core/error/app_exception.dart';
import 'package:lol_champion/features/champion/data/datasource/champion_remote_data_source.dart';
import 'package:lol_champion/features/champion/domain/entity/champion.dart';
import 'package:lol_champion/features/champion/domain/repository/champion_repository.dart';
import 'package:lol_champion/features/champion/domain/repository/patch_version_repository.dart';

class ChampionRepositoryImpl implements ChampionRepository {
  const ChampionRepositoryImpl(this._remote, this._patchVersionRepository);

  final ChampionRemoteDataSource _remote;
  final PatchVersionRepository _patchVersionRepository;

  @override
  Future<List<Champion>> getChampionList() async {
    try {
      final version = await _patchVersionRepository.getPatchVersion();
      final response = await _remote.getChampionList(version);
      final champions =
          response.data.values.map((champion) => champion.toEntity()).toList()
            ..sort((a, b) => a.name.compareTo(b.name));
      return List.unmodifiable(champions);
    } on DioException catch (error) {
      throw NetworkException(error);
    } on AppException {
      rethrow;
    } on Object catch (error) {
      throw DataException(error);
    }
  }

  @override
  Future<Champion> getChampion(String championId) async {
    try {
      final version = await _patchVersionRepository.getPatchVersion();
      final response = await _remote.getChampion(version, championId);
      final champion = response.data[championId];
      if (champion == null) {
        throw ChampionNotFoundException(championId);
      }
      return champion.toEntity();
    } on DioException catch (error) {
      throw NetworkException(error);
    } on AppException {
      rethrow;
    } on Object catch (error) {
      throw DataException(error);
    }
  }
}
